docs/docs/creating-and-modifying-pages.md

```diff
@@ -92,13 +92,103 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
+
+In the initial approach you have seen how the `gatsby-node.js` file would have a query block like so :
+
+```graphql
+  const queryResults = await graphql(`
+    query AllProducts {
+      AllProducts {
+        nodes {
+          id
+        }
+      }
+    }
+  `);
+```
+
+Using the `id` as an access point to query for other properties in the template is the default approach. However, Suppose you had a list of products with properties you would like to query for. Handling the query all from `gatsby-node.js` the query would look like this:
+
+```javascript:tittle=gatsby-node.js
+
+exports.createPages = async ({ graphql, actions }) => {
+  const { createPage } = actions;
+  const queryResults = await graphql(`
+    query AllProducts {
+      AllProducts {
+        nodes {
+          id
+          name
+          price
+          description
+        }
+      }
+    }
+  `);
+
+  const productTemplate = path.resolve(`src/templates/product.js`);
+  queryResults.data.allProducts.nodes.forEach(node => {
+    createPage({
+      path: `/products/${node.id}`,
+      component: productTemplate,
+      context: {
+        // This time the entire product is passed down as context
+        product
+      }
+    });
+  });
+};
+};
+```
+
+> You are now requesting all the data you need in a single query (this requires server-side support to fetch many products in a single database query).
+
+> As long as you can pass this data down to the template component via `pageContext`, there is no need for the template to make a GraphQL query at all.
+
+Your template file would look like this:
+
+```javascript:tittle=src/templates/product.js
+function Product({ pageContext }) {
+  return (
+    <div>
+      Name: {pageContext.name}
+      Price: {pageContext.price}
+      Description: {pageContext.description}
+    </div>
+  )
+}
+```
+
+### Pefromance implecations
+
+Using the PageContext props in the template component can come with its performance advantages, of getting in all the data you need at build time; from the createPages API. This prevents having a graphql query in the template component.
+
+It does come with the advantage of querying your data from one place after declaring the `context` parameter.
+
+However, it doesn’t give you the opportunity to know what exactly you are querying for in the template and if any changes occur in the component query structure in gatsby-node.js. Hot reload is taken off the table and the site needs to be rebuilt for changes to reflect.
+
+Gatsby stores page metadata (including context) in a redux store (which also means that it stores its memory). For large enough sites (either number of pages and/or amount of data that is being passed via page context) this will cause problems. This might be out of memory crashes if it's too much data or degraded performance.
```

Comment:
I'm not sure I fully understand this paragraph 🤔 

> in a redux store (which also means that it stores its memory)

Is the redux store also stored in-memory?

> (either number of pages and/or amount of data that is being passed via page context)

Does this mean something like based on the number of pages or data? Maybe saying "For larger sites..." because a large site is relative. What makes a site too big?

Reply:
I’ll rephrase this, also I pinged Pieh about this classification of “Larger site” . I think in this case maybe a prompt to say the pages are becoming too much would help ?

Reply:
I rephrased. 😄